### PR TITLE
demux_lavf: try `handler_name` as title fallback

### LINKED ISSUE #8488: mpv don't show audio tracks' titles in mp4 files
### Important Information

- mpv version: v0.33.0
- macOS Version: 10.14.6
- [Source of the mpv binary](https://laboratory.stolendata.net/~djinn/mpv_osx/)

### Reproduction steps

Open the sample file, then call Playback - Show tracks.
I haven't a config file.

### Expected behavior
Showing audio track's titles like in VLC, MediaInfo.

### Actual behavior
mpv don't show audio tracks' titles in mp4 files.

What I see in mpv:

And what I see in MediaInfo (I've set off track titles — MVO and ORT):
изображение

VLC also shows track titles:
изображение

### Log file

[Link](https://0x0.st/-iL_.txt)

### Sample files

[Sample](https://0x0.st/-zjq.m4v)

### PR DESCRIPTION
**Pull request by gqgs**
Fixes #8488

### CODE REVIEW

File: demux/demux_lavf.c [hunk 1]
```diff
@@ -800,6 +800,14 @@ static void handle_new_stream(demuxer_t *demuxer, int i)
         AVDictionaryEntry *title = av_dict_get(st->metadata, "title", NULL, 0);
         if (title && title->value)
             sh->title = talloc_strdup(sh, title->value);
+        else {
+            AVDictionaryEntry *handler_name = av_dict_get(st->metadata, "handler_name", NULL, 0);
+            if (handler_name && handler_name->value &&
+                strcmp(handler_name->value, "VideoHandler")&&
```
Comment (LaserEyess) [inline demux/demux_lavf.c:806]: nit, but there should be a space between the `&&`

**Comment by LaserEyess** [comment]
I think this looks good but can you document where the "*Handler" strings come from? Best I could find was [this](https://stackoverflow.com/questions/27518432/how-to-set-custom-handler-name-metadata-for-subtitle-stream-using-ffmpeg) which looks ffmpeg specific.

**Comment by gqgs** [comment]
@LaserEyess Yes it does seem to be very ffmpeg specific. More specifically their [MOV encoder](https://github.com/FFmpeg/FFmpeg/blob/9b4b0df47009ebc3d26bf3f1b1eff668fd028cc6/libavformat/movenc.c#L2773) is the only place I've seen this *Handler names being defined.

It uses these generic names by default or something more descriptive if the [stream provides an alternate description in the metadata](https://github.com/FFmpeg/FFmpeg/blob/9b4b0df47009ebc3d26bf3f1b1eff668fd028cc6/libavformat/movenc.c#L2810). 

Should we not bother filtering these generic titles in this case?

**Comment by avih** [comment]
Isn't this an ffmpeg issue/bug that it doesn't put the name at the "title" entry? Is it documented some place that one should look at the "handler_name" entry?

**Comment by gqgs** [comment]
@avih The relevant specs here seem to be ISO/IEC 14496-12 8.4.3.3 and [QTFF](https://developer.apple.com/library/archive/documentation/QuickTime/QTFF/QTFFChap2/qtff2.html#//apple_ref/doc/uid/TP40000939-CH204-33004). The semantic specification doesn't say much other than it __should__ be valid human‐readable UTF-8 and that this field __can__ be used for "inspection purposes". As for the *Handler values, it seems to be an unilateral decision made by the ffmpeg devs to use those specific values by default. They are not defined in any specification as far as I can tell.

**Comment by avih** [comment]
> The relevant specs here seem to be ISO/IEC 14496-12 8.4.3.3 and QTFF...

As far as I can tell, it specifies an `hdlr` mov atom type - which is a "handler reference". Is it documented for ffmpeg that `hdlr` maps to "handler_name"?

Shouldn't ffmpeg map the specifics of the container to an abstract generic thing which clients can use to get the track name/title without caring that it's a `mov` container?

What about other containers? Do they also need custom ways to access a track title value?

> can be used for "inspection purposes"

To be honest, I don't know if this qualifies as track name. To me it sounds more like an internal thing. Not saying it's useless, but it doesn't sound to me like something which is intended to be visible to users.

Without knowing much about ffmpeg APIs, overall, to me this sounds like an issue at the ffmpeg decoder. I think that if it decides that something at the container qualifies as a track title, then it should expose it as a generic title for the track.

But maybe I'm looking at it wrongly? Maybe @jeeb knows more?

**Comment by avih** [comment]
Also, note that your patch now prevents classifying a mov clip track as "visual impaired" or "hearing impaired", because you claim the field is mandatory, and so it will always exist at `handler_name`, which will always override the ..impaired detection.

**Comment by jeeb** [comment]
Sorry for getting to this a bit late.

> Also, note that your patch now prevents classifying a mov clip track as "visual impaired" or "hearing impaired", because you claim the field is mandatory, and so it will always exist at `handler_name`, which will always override the ..impaired detection.

That only consists of the stream title, if you look at the block just before it does set the `visual_impaired_track` flag for the `sh_stream`. So the classification is correct, but it will just not show up as part of the stream title. This could be switched to an append for fall-back titles :) .

Something like the following (E_NOT_SURE about the `talloc_strdup_append` usage, but otherwise should give an example of how it could be handled).
```diff
diff --git a/demux/demux_lavf.c b/demux/demux_lavf.c
index 71fb9fc289..73af82cf57 100644
--- a/demux/demux_lavf.c
+++ b/demux/demux_lavf.c
@@ -774,6 +774,8 @@ static void handle_new_stream(demuxer_t *demuxer, int i)
     MP_TARRAY_APPEND(priv, priv->streams, priv->num_streams, info);
 
     if (sh) {
+        bool proper_title_set = false;
+
         sh->ff_index = st->index;
         sh->codec->codec = mp_codec_from_av_codec_id(codec->codec_id);
         sh->codec->codec_tag = codec->codec_tag;
@@ -797,13 +799,25 @@ static void handle_new_stream(demuxer_t *demuxer, int i)
             sh->still_image = true;
         if (priv->format_hack.use_stream_ids)
             sh->demuxer_id = st->id;
+
         AVDictionaryEntry *title = av_dict_get(st->metadata, "title", NULL, 0);
-        if (title && title->value)
+        if (title && title->value) {
             sh->title = talloc_strdup(sh, title->value);
-        if (!sh->title && st->disposition & AV_DISPOSITION_VISUAL_IMPAIRED)
-            sh->title = talloc_asprintf(sh, "visual impaired");
-        if (!sh->title && st->disposition & AV_DISPOSITION_HEARING_IMPAIRED)
-            sh->title = talloc_asprintf(sh, "hearing impaired");
+            proper_title_set = true;
+        } else {
+            // fallbacks
+        }
+
+        if (!proper_title_set && st->disposition & AV_DISPOSITION_VISUAL_IMPAIRED)
+            sh->title = sh->title ?
+                        talloc_strdup_append(sh->title, " (visual impaired)") :
+                        talloc_asprintf(sh, "visual impaired");
+
+        if (!proper_title_set && st->disposition & AV_DISPOSITION_HEARING_IMPAIRED)
+            sh->title = sh->title ?
+                        talloc_strdup_append(sh->title, " (hearing impaired)") :
+                        talloc_asprintf(sh, "hearing impaired");
+
         AVDictionaryEntry *lang = av_dict_get(st->metadata, "language", NULL, 0);
         if (lang && lang->value && strcmp(lang->value, "und") != 0)
             sh->lang = talloc_strdup(sh, lang->value);

```

Now, back to utilizing the handler identifier as a fall-back title:

My base opinion is that I do not like utilizing this value, as it does not seem to have been meant for such usage. I do not think that the `handler_name` can be utilized well as a _useful_ value in most of the cases. Some files might have something written into it, but it most definitely is not generally meant to be a "track title". The value it is describing is basically the _track type_ (video, audio, text, hint etc).

From the specification (14496-12):
```
aligned(8) class HandlerBox extends FullBox(‘hdlr’, version = 0, 0) { 
  unsigned int(32)  pre_defined = 0; 
  unsigned int(32)  handler_type; 
  const unsigned int(32)[3]  reserved = 0; 
  string  name; 
}
```

(the `handler_type` is basically a fourcc for a type of stream, such as `text` for certain subtitles)

> _name_ is a null‐terminated string in UTF‐8 characters which gives a human readable name for the track type (for debugging and inspection purposes).

And to give some examples of various values written by various writers:
* QuickTime (QTFF ref. implementation more or less) utilizes `Core Media Video/Audio` etc
* Google writes `ISO Media file produced by Google Inc. Created on: 04/10/2020.` even though the field is track (stream) related and not global media stuff.
* L-SMASH also utilizes `XYZ Handler`.
* Stagefright (or whatever else Android is utilizing atm) uses `VideoHandle`/`AudioHandle` etc
* GPAC has had various strings throughout the years, and generally has the string `GPAC` in there somewhere.

So uh, with a quick look if one would want to filter things out:
```c
AVDictionaryEntry *handler_name = av_dict_get(st->metadata, "handler_name", NULL, 0);
if (handler_name && handler_name->value &&
    // Android, FFmpeg and L-SMASH utilize XYZHandle(r)
    !strstr(handler_name->value, "Handle") &&
    // Youtube flags its own muxes with this
    !strstr(handler_name->value, "Google Inc") &&
    // QuickTime flags its with Core Media XYZ
    !strstr(handler_name->value, "Core Media") &&
    // GPAC writes its project name in handler
    !strstr(handler_name->value, "GPAC"))
    sh->title = talloc_strdup(sh, handler_name->value);
```

I don't think I need to say why I'm not a fan of such :) . Apparently for a while Handbrake did utilize the handler name for track titles, not sure if it is utilizing it any more (the sample was generated with `HandBrake 1.3.3 2020061300`).

Thankfully, the MP4 sample seems to also contain some user data at least under one of the audio tracks having a title
```
            [udta: User Data Box]
                position = 3991500
                size = 71
                [name]
                    position = 3991508
                    size = 28
                [titl]
                    position = 3991536
                    size = 35
```
(udta being an actual metadata box)

And funny enough, both of those two boxes have the same track title value there :D . I will be checking on FFmpeg's side why neither of those gets utilized (as well as double-checking QTFF's definition of those two).

**Pull request closed without merging**